### PR TITLE
Fix instr Function to Correctly Handle Unicode Characters in Arrow DataFusion

### LINKED ISSUE #9026: instr function doesn't correctly account for non-ascii strings
### Describe the bug

the instr function doesn't correctly return the position of the matched string, rather it returns the position of the matched byte index. The issue is in this code in string_expressions.rs
```
.map(|(string, substr)| match (string, substr) {
                    (Some(string), Some(substr)) => string
                        .find(substr)
                        .map_or(Some(0), |index| Some((index + 1) as i32)),
                    _ => None,
                })
```
See the documentation for the find function.

The expected result from the end users perspective is not the byte index but rather the index of the first matching unicode grapheme in the string.

### To Reproduce

❯ docker run -it -v /tmp:/data datafusion-cli
DataFusion CLI v35.0.0
❯ select instr('实现这个函数', '函')
;
+----------------------------------------+
| instr(Utf8("实现这个函数"),Utf8("函")) |
+----------------------------------------+
| 13                                     |
+----------------------------------------+
1 row in set. Query took 0.002 seconds.

5 should be the correct position.

### Expected behavior

The returned position should be the position of the first matching unicode grapheme in the string, not the first matching byte index.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by zy-kkk**
## Which issue does this PR close?

Closes #9026.

### CODE REVIEW

File: datafusion/physical-expr/src/string_expressions.rs [hunk 1]
```diff
@@ -303,15 +303,29 @@ pub fn instr<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
     let string_array = as_generic_string_array::<T>(&args[0])?;
     let substr_array = as_generic_string_array::<T>(&args[1])?;
 
+    fn process_instr(string: &str, substr: &str) -> Option<i64> {
```
Comment (alamb) [inline datafusion/physical-expr/src/string_expressions.rs:306]: this is likely substantially slower than the previous implementation as it copies the strings into Vecs on each row

File: datafusion/physical-expr/src/string_expressions.rs [hunk 2]
```diff
@@ -303,15 +303,29 @@ pub fn instr<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
...
     match args[0].data_type() {
         DataType::Utf8 => {
             let result = string_array
                 .iter()
                 .zip(substr_array.iter())
-                .map(|(string, substr)| match (string, substr) {
-                    (Some(string), Some(substr)) => string
-                        .find(substr)
-                        .map_or(Some(0), |index| Some((index + 1) as i32)),
+                .map(|(string_opt, substr_opt)| match (string_opt, substr_opt) {
```
Comment (alamb) [inline datafusion/physical-expr/src/string_expressions.rs:325]: I wonder if we could use `char_indices` here to find the relevant offset with the matched string? https://doc.rust-lang.org/std/primitive.str.html#method.char_indices

🤔 

If not, we can probably at least special case on `is_ascii`

**Review by alamb — APPROVED** [review]
Thank you for this contribution @zy-kkk 

I think this does a nice job of fixing the functionality, but I think we should think about performance a bit if possible

cc @Omega359

**Comment by Omega359** [comment]
Unless I'm incorrect the instr function and the existing strpos function are logically identical but functionally different - I am pretty sure that strpos correctly handles unicode characters (not graphemes). I think that this function should be removed and replaced with an alias on strpos and the docs updated.

**Comment by alamb** [comment]
> Unless I'm incorrect the instr function and the existing strpos function are logically identical but functionally different - I am pretty sure that strpos correctly handles unicode characters (not graphemes). I think that this function should be removed and replaced with an alias on strpos and the docs updated.

That sounds like a much better solution to me.

**Comment by alamb** [comment]
I don't think this PR is waiting on feedback anymore so marking it as draft. Please mark as ready for review when it is

**Comment by Omega359** [comment]
With the removal of instr in #9391 this issue should no longer be valid and can be closed. @alamb

**Comment by alamb** [comment]
Please reopen this @zy-kkk  if you disagree or if there is something we missed

**Comment by zy-kkk** [comment]
> Please reopen this @zy-kkk if you disagree or if there is something we missed

I have no problem with remove instr code, add instr as an alias for strpos, thanks for your work

**Pull request closed without merging**